### PR TITLE
vim: Expose the search-by-type feature

### PR DESCRIPTION
**Pull request by Julow**
This adds the `:MerlinSearchType` command to query for a value with a matching type. It works like the previous `:MerlinSearch` but uses the new search feature added in https://github.com/ocaml/merlin/pull/1828, the search results are shown in the completion menu.

`:MerlinSearch` switches between polarity search and search by type depending on the first character of the query, like is done in the emacs plugin.

Documentation about the three search functions is added.

### CODE REVIEW

**Review by xvw — APPROVED** [review]
Thanks a lot!

File: vim/merlin/autoload/merlin.py [hunk 1]
```diff
@@ -365,15 +365,15 @@ def command_holes():
...
-    return re.sub(re_spaces_around_nl, "\n", re.sub(re_spaces, " ", str)).replace("'", "''").strip()
-
+# Turns Merlin's search-by-polarity or complete-prefix entries into a Vim's completion-item list
 def vim_fillentries(entries, vimvar):
-    prep = vim_complete_prepare
-    prep_nl = vim_complete_prepare_preserve_newlines
     for prop in entries:
-        vim.command("let tmp = {'word':'%s','menu':'%s','info':'%s','kind':'%s'}" %
-                (prep(prop['name']),prep(prop['desc']),prep_nl(prop['info']),prep(prop['kind'][:1])))
+        vim.command("let tmp = " + vim_record({
```
Comment (voodoos) [inline vim/merlin/autoload/merlin.py:371]: It's unclear to me that vim_record behave similarly as the previous  use of `prep` and `prep_nl`. Does it "preserve newlines" ? And only for the "info" field ?

File: vim/merlin/autoload/merlin.vim [hunk 2]
```diff
@@ -284,15 +284,33 @@ function! merlin#PolaritySearch(debug,query)
...
+
+" Do a polarity search or a search by type depending on the first character of
+" the query.
+function! merlin#Search(debug, query)
+  if a:query =~ "^[-+]"
+    call merlin#PolaritySearch(a:debug, a:query)
+  else
+    call merlin#SearchByType(a:debug, a:query)
+  endif
+endfunction
```
Comment (voodoos) [inline vim/merlin/autoload/merlin.vim:312]: @xvw this detection is also made on Merlin side isn't it ?
Maybe we could have a simpler interface in both emacs and vim:

- Keep the existing PolaritySearch
- Add a new SearchByType or Search, that does indeed also support the polarity search syntax.

**Review by voodoos — COMMENTED** [review]
Thanks @Julow !

File: vim/merlin/autoload/merlin.vim (hunk 2 above)
Comment (xvw) [reply vim/merlin/autoload/merlin.vim:312]: Yes, in Emacs I've moved the original `search` behaviour into a `search-by-polarity` command and then the `search` command relies on the merlin `search-by-type` command (which classifies the query)

File: vim/merlin/autoload/merlin.py (hunk 1 above)
Comment (Julow) [reply vim/merlin/autoload/merlin.py:372]: You are right, there was a problem here!
I've fixed `vim_record` to properly encode newlines and added back the `prep` and `prep_nl` functions. I'm not sure in which cases these are useful.
I've just learnt from this "info" feature and I find it awesome. I tested that completion works fine with these settings:
```
let g:merlin_completion_with_doc = 1
set completeopt+=popup
```

**Comment by voodoos** [comment]
Oh, I completely forgot about this one, sorry. @Julow is it ready ? 🙂

**Comment by voodoos** [comment]
Thanks !

**Pull request merged**